### ISSUE TITLE
removal of service-proxy label on Endpoints is not reconciled

### ISSUE THREAD

**Issue by maplain** [kubernetes/kubernetes#98577]
**What happened**:
When the **service.kubernetes.io/service-proxy-name** label is removed from an Endpoints object, kube-proxy doesn't reconcile the service. Therefore, the Service stays unavailable.

**What you expected to happen**:
kube-proxy reconciles the Endpoints, removes the "no endpoints" iptable rule, and installs correct rules to enable accessbility.

**How to reproduce it (as minimally and precisely as possible)**:
1. Check if we have a clean test environment
```
➜ ✗ kubectl get service
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   100.64.0.1   <none>        443/TCP   34h
➜  ✗ kubectl get ep
NAME         ENDPOINTS            AGE
kubernetes   10.92.124.198:6443   34h
➜  ✗ kubectl get pod
No resources found in default namespace.
```
2. creates a deployment
```
➜  ✗ cat << EOF | kubectl apply -f -
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-proxy
spec:
  replicas: 2
  selector:
    matchLabels:
      app: test-proxy
  template:
    metadata:
      labels:
        app: test-proxy
    spec:
      serviceAccountName: default
      containers:
        - name: nginx
          image: xxx/nginx:latest

EOF
deployment.apps/test-proxy created
```
3. Find the Pod information
```
➜ ✗ kubectl get pod -o wide
NAME                          READY   STATUS    RESTARTS   AGE     IP            NODE                       NOMINATED NODE   READINESS GATES
test-proxy-67db64d8cc-pg2qn   1/1     Running   0          2m45s   100.96.1.12   w1-md-0-86c6b7b994-r7jrz   <none>           <none>
test-proxy-67db64d8cc-r72x8   1/1     Running   0          2m45s   100.96.1.13   w1-md-0-86c6b7b994-r7jrz   <none>           <none>

➜ ✗ kubectl get pod test-proxy-67db64d8cc-pg2qn -o json | jq -cr '.metadata.resourceVersion'
500728
➜ ✗ kubectl get pod test-proxy-67db64d8cc-pg2qn -o json | jq -cr '.metadata.uid'
fe3fefaf-0f04-4b49-904a-20566af788ec

➜  ✗ kubectl get pod test-proxy-67db64d8cc-r72x8 -o json | jq -cr '.metadata.resourceVersion'
500719
➜  ✗ kubectl get pod test-proxy-67db64d8cc-r72x8 -o json | jq -cr '.metadata.uid'
48a21c5c-5144-4726-ba30-5f55977c7860
```
4. Manually put together an Endpoints and create it
```
➜ ✗ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-proxy
  namespace: default
subsets:
  - addresses:
      - ip: 100.96.1.12
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pg2qn
          namespace: default
          resourceVersion: "500728"
          uid: fe3fefaf-0f04-4b49-904a-20566af788ec
      - ip: 100.96.1.13
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pqrnp
          namespace: default
          resourceVersion: "500719"
          uid: 48a21c5c-5144-4726-ba30-5f55977c7860
    ports:
      - port: 80
        protocol: TCP
EOF
endpoints/test-proxy created
```

5. Create a LB type SVC for the Endpoints
```
➜  ✗ cat << EOF  | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-proxy
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  type: LoadBalancer
EOF
service/test-proxy created
```

6. Verify it's accessible
```
➜  ✗ kubectl get service
NAME         TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)        AGE
kubernetes   ClusterIP      100.64.0.1      <none>         443/TCP        35h
test-proxy   LoadBalancer   100.69.222.79   10.92.96.254   80:31280/TCP   43s
➜  ✗ curl 10.92.96.254
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
➜  ✗ curl 10.92.96.254
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

7. delete the Endpoints and manually create a new one with label "service.kubernetes.io/service-proxy-name" applied:
```
➜  ✗ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-proxy
  namespace: default
  labels:
    service.kubernetes.io/service-proxy-name: others
subsets:
  - addresses:
      - ip: 100.96.1.12
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pg2qn
          namespace: default
          resourceVersion: "500728"
          uid: fe3fefaf-0f04-4b49-904a-20566af788ec
      - ip: 100.96.1.13
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pqrnp
          namespace: default
          resourceVersion: "500719"
          uid: 48a21c5c-5144-4726-ba30-5f55977c7860
    ports:
      - port: 80
        protocol: TCP
EOF
endpoints/test-proxy created
```
8. Verify the Service is inaccessible
```
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
```
9. Edit the Endpoints to remove the label
```
➜ ✗ kubectl edit ep test-proxy -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Endpoints","metadata":{"annotations":{},"labels":{"service.kubernetes.io/service-proxy-name":"others"},"name":"test-proxy","namespace":"default"},"subsets":[{"addresses":[{"ip":"100.96.1.12","nodeName":"w1-md-0-86c6b7b994-r7jrz","targetRef":{"kind":"Pod","name":"test-proxy-67db64d8cc-pg2qn","namespace":"default","resourceVersion":"500728","uid":"fe3fefaf-0f04-4b49-904a-20566af788ec"}},{"ip":"100.96.1.13","nodeName":"w1-md-0-86c6b7b994-r7jrz","targetRef":{"kind":"Pod","name":"test-proxy-67db64d8cc-pqrnp","namespace":"default","resourceVersion":"500719","uid":"48a21c5c-5144-4726-ba30-5f55977c7860"}}],"ports":[{"port":80,"protocol":"TCP"}]}]}
  creationTimestamp: "2021-01-29T18:28:40Z"
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
      f:subsets: {}
    manager: kubectl-client-side-apply
    operation: Update
    time: "2021-01-29T18:28:40Z"
  name: test-proxy
  namespace: default
  resourceVersion: "502865"
  selfLink: /api/v1/namespaces/default/endpoints/test-proxy
  uid: 2a600292-97d6-43d9-92b7-2f2e280c1ab2
subsets:
- addresses:
  - ip: 100.96.1.12
    nodeName: w1-md-0-86c6b7b994-r7jrz
    targetRef:
      kind: Pod
      name: test-proxy-67db64d8cc-pg2qn
      namespace: default
      resourceVersion: "500728"
      uid: fe3fefaf-0f04-4b49-904a-20566af788ec
  - ip: 100.96.1.13
    nodeName: w1-md-0-86c6b7b994-r7jrz
    targetRef:
      kind: Pod
      name: test-proxy-67db64d8cc-pqrnp
      namespace: default
      resourceVersion: "500719"
      uid: 48a21c5c-5144-4726-ba30-5f55977c7860
  ports:
  - port: 80
    protocol: TCP
```

10. Describe Service, Endpoints, and query the Service, which is still inaccessible
```
➜ ✗ kubectl describe ep test-proxy
Name:         test-proxy
Namespace:    default
Labels:       <none>
Annotations:  <none>
Subsets:
  Addresses:          100.96.1.12,100.96.1.13
  NotReadyAddresses:  <none>
  Ports:
    Name     Port  Protocol
    ----     ----  --------
    <unset>  80    TCP

Events:  <none>
➜ ✗ kubectl describe service test-proxy
Name:                     test-proxy
Namespace:                default
Labels:                   <none>
Annotations:              <none>
Selector:                 <none>
Type:                     LoadBalancer
IP:                       100.69.222.79
LoadBalancer Ingress:     10.92.96.254
Port:                     <unset>  80/TCP
TargetPort:               80/TCP
NodePort:                 <unset>  31280/TCP
Endpoints:                100.96.1.12:80,100.96.1.13:80
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
```

11. check the iptables rules on the node
```
# iptables-save | grep 10.92.96.254
-A KUBE-SERVICES -d 10.92.96.254/32 -p tcp -m comment --comment "default/test-proxy has no endpoints" -m tcp --dport 80 -j REJECT --reject-with icmp-port-unreachable
```
12. Manually kill kube-proxy pod to restart
```
➜ ✗ kubectl get pods -n kube-system
NAME                                             READY   STATUS    RESTARTS   AGE
...
kube-proxy-hr2fj                                 1/1     Running   0          35h
kube-proxy-jw7st                                 1/1     Running   0          35h
...
➜ ✗ kubectl delete pod -n kube-system kube-proxy-hr2fj
pod "kube-proxy-hr2fj" deleted
➜  ✗ kubectl delete pod -n kube-system kube-proxy-jw7st
pod "kube-proxy-jw7st" deleted
➜ ✗ kubectl get pods -n kube-system
...
kube-proxy-6bbs8                                 1/1     Running   0          25s
kube-proxy-qtspc                                 1/1     Running   0          6s
...
```
13. Service is still inaccessible and iptable rule stays unchanged
```
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
➜ ✗ curl --max-time 1 10.92.96.254
curl: (7) Failed to connect to 10.92.96.254 port 80: Connection refused
...
# iptables-save | grep 10.92.96.254
-A KUBE-SERVICES -d 10.92.96.254/32 -p tcp -m comment --comment "default/test-proxy has no endpoints" -m tcp --dport 80 -j REJECT --reject-with icmp-port-unreachable
```
14. Delete the Endpoint and recreate it without the label
```
➜  ✗ kubectl delete ep test-proxy
endpoints "test-proxy" deleted
➜ ✗ cat << EOF | kubectl apply -f -
pipe heredoc> apiVersion: v1
kind: Endpoints
metadata:
  name: test-proxy
  namespace: default
subsets:
  - addresses:
      - ip: 100.96.1.12
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pg2qn
          namespace: default
          resourceVersion: "500728"
          uid: fe3fefaf-0f04-4b49-904a-20566af788ec
      - ip: 100.96.1.13
        nodeName: w1-md-0-86c6b7b994-r7jrz
        targetRef:
          kind: Pod
          name: test-proxy-67db64d8cc-pqrnp
          namespace: default
          resourceVersion: "500719"
          uid: 48a21c5c-5144-4726-ba30-5f55977c7860
    ports:
      - port: 80
        protocol: TCP
pipe heredoc> EOF
endpoints/test-proxy created
```
15. Service is back to normal
```
➜ ✗ curl --max-time 1 10.92.96.254
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

**Anything else we need to know?**:
no

**Environment**:
- Kubernetes version (use `kubectl version`):
```
➜ ✗ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T13:41:02Z", GoVersion:"go1.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3+vmware.1", GitCommit:"2ac9e7ea06a1230ca196931def19d2bb67b580c7", GitTreeState:"clean", BuildDate:"2020-10-30T07:25:38Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```
# cat /etc/os-release
NAME="VMware Photon OS"
VERSION="3.0"
ID=photon
VERSION_ID=3.0
PRETTY_NAME="VMware Photon OS/Linux"
ANSI_COLOR="1;34"
HOME_URL="https://vmware.github.io/photon/"
BUG_REPORT_URL="https://github.com/vmware/photon/issues"
```
- Kernel (e.g. `uname -a`):
```
# uname -a
Linux w1-md-0-86c6b7b994-r7jrz 4.19.150-1.ph3 #1-photon SMP Fri Oct 23 02:29:37 UTC 2020 x86_64 GNU/Linux
```
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by jayunit100** [comment]
I'll attempt to reproduce this today , especially since your using tanzu :) (we noticed the photon in your bug report) 
/assign

**Comment by jayunit100** [comment]
ok so i didnt get to fully repro but, 
- great bug report , mostly makes sense.  I got stuck on the `Loadbalancer` part though because i was testing in Kind.  To reproduce this - do you need like an NSX loadbalancer (i see that 100 IP address)... or does this happen w/ literally any cloud loadbalancer? 
- gotta get rid of those `heredocs` i think they were accidental pastes ? 

Thanks again ! Sorry i only had like 15 mintues to start repro'ing but so far it looks like a great bug report.  Ill get back to it later this wk, unless someone beats me to it.

**Comment by jayunit100** [comment]
from zoom today `jay check in the use that kube=proxy is using endpoints not slices` (@aojea)

**Comment by aojea** [comment]
also, the label `service.kubernetes.io/service-proxy-name` is meant to be used in the Service object, not in the endpoints, because the endpoints are autogenerated by the controller, and then they can be overwrite the user label.

**Comment by njuptlzf** [comment]
Hei @aojea.  Before you reply, I thought about adding the `UpdateFunc ` to fix this problem, am I wrong?

**Comment by jayunit100** [comment]
@aojea , makes sense what you said but now....  im more confused. `service-proxy-name` when applied to an endpoint should  either.... 
 
1) Be IGNORED ENTIRELY by the kube proxy  OR
2) be reconciled so that removal of the label is the same behaviour as if the label NEVER existed

So its *still* a bug , right ?

**Comment by aojea** [comment]
Things to take into account:
1. the kube-proxy informer filters everything with `service-proxy-name`  , that means endpoints are services

https://github.com/kubernetes/kubernetes/blob/6dc0047396d3aad16928d346da8a5052d8a824fa/cmd/kube-proxy/app/server.go#L726-L730

2. the endpoints controller does´t process updates on Endpoints, and is not likely that is going to change

https://github.com/kubernetes/kubernetes/issues/98066#issuecomment-763920732
https://github.com/kubernetes/kubernetes/pull/98122#issuecomment-766353557

3. the endpoints controller autogenerate endpoints for services with a selector, that means that when you create or delete (not update as explained in 2. ) custom endpoints you are going to race with it

The key is the step 9., 

> Edit the Endpoints to remove the label

That operation, edit to remove the label IIRC, is going to be a PATCH to the Endpoint object, that patch seems to not be processed by kube-proxy, and my apimachinery fu is not strong enough to know if it should be processed or not per the kube-proxy config in 1.

I wish I have more time to hack on this, this will be easily reproduced with KIND, you don´t need a loadbalancer service, a clusterip or nodeport one will be enough.

Just enable enough logging in kube-proxy and the controller-manager so you can correlate your actions with what the endpoints controller and the kube-proxy are doing, if the PATCH is not processed by kube-proxy and that is how the informer is supposed to work, this is working as expected.

**Comment by jayunit100** [comment]
AHHHhhh ok so the steps here are:
- determine WTH should happen when you PATCH the EP object , based on the controller in kubernetes/cmd/kube-proxy/app/server.go.go
- determine wether we're doing ^, or if there are corner cases where we don't do it properly, or if its normal to never respond to PATCH operations against the endpoints object.

@maplain now that the context is clear here - do you want to explore that ?   or should we find someone else to look into this ?    

...  Im not a kube-proxy god so i can't say for sure but the TLDR for me here is  - Endpoints are complicated, and editing them manually, unless you really know what your doing  is probably a bad idea :).  @ the very least we really should document this thoroughly as i bet other folks will run into this to .

**Comment by maplain** [comment]
thx @aojea and @jayunit100 for taking a look.

If the label is supposed to be used for Service only, why would creating an Endpoints with that label make the Service unavailable?

**Comment by aojea** [comment]
> 1. the kube-proxy informer filters everything with the `service-proxy-name`  label, that means endpoints and services

**Comment by maplain** [comment]
gotcha. thx @aojea!

So what happened is:
1. kube-proxy doesn't watch Service or Endpoints with `service-proxy-name` label. Therefore the Endpoint created with that label is not reconciled at all, which leads to the REJECT rule;
2. PATCH on Endpoint is not processed, so updating it to remove the label doesn't take effect either.

This also explains why the following flow works as expected:
1. create Service and an Endpoint without the `service-proxy-name` label. Service works as expected;
2. Add the label to both, then Service is inaccessible;
3. Remove the label from both, Service is back to normal

Especially the label removal in step3. It's not actually the PATCH that works, but really from kube-proxy's perspective, the Endpoint has never had that label applied.

Then here is my question:
1. why does kube-proxy filter Endpoint with `service-proxy-name` label? It looks like no matter the Service has a selector or not, filtering Service is enough to achieve what we what.

**Comment by aojea** [comment]
service labels are mirrored to the endpoint
https://github.com/kubernetes/kubernetes/blob/3f09686958f00420ae1608ae9997ceb4d06e7006/pkg/controller/endpoint/endpoints_controller.go#L519

> 1. why does kube-proxy filter Endpoint with `service-proxy-name` label? It looks like no matter the Service has a selector or not, filtering Service is enough to achieve what we what.

IIRC the filtering is on the apiserver, so you don't receive objects that you don't need to the nodes

**Comment by maplain** [comment]
> service labels are mirrored to the endpoint
should it do so even if the Service is selectorless and Endpoint is actually managed by users?

**Comment by aojea** [comment]
> should it do so even when the Service is selectorless and Endpoint is actually managed by users?

that is not always true, the Endpoints are autogenerated for Services with selectors, in this case you are trying to do something that should not be allowed, because you are modified an Endpoint that belongs to the Endpoints controller.

**Comment by thockin** [comment]
If you run this on an endpointslice-enabled proxy, the removal of the label from Endpoints SHOULD propagate to EndpointSlice.  Can we confirm that?  If not, we should fix that.

**Comment by maplain** [comment]
What I did is:
1. create a selector Service and an Endpoint with two Pods manually populated;
2. The Endpoint has the label `service.kubernetes.io/service-proxy-name: others` applied;
```
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Endpoints","metadata":{"annotations":{},"labels":{"service.kubernetes.io/service-proxy-name":"others"},"name":"loadbalancer","namespace":"default"},"subsets":[{"addresses":[{"ip":"100.96.1.10","nodeName":"w1-md-0-54cbf578bf-nt4d8","targetRef":{"kind":"Pod","name":"loadbalancer-78df87b854-8t844","namespace":"default","resourceVersion":"1752571","uid":"388e0bad-abb0-4f47-870a-2ad1f01a428e"}},{"ip":"100.96.1.11","nodeName":"w1-md-0-54cbf578bf-nt4d8","targetRef":{"kind":"Pod","name":"loadbalancer-78df87b854-nkpcl","namespace":"default","resourceVersion":"1752573","uid":"46cefe5a-efe8-434a-8209-2eafd717e4ee"}}],"ports":[{"port":80,"protocol":"TCP"}]}]}
  creationTimestamp: "2021-03-18T23:29:27Z"
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
      f:subsets: {}
    manager: kubectl-client-side-apply
    operation: Update
    time: "2021-03-18T23:29:27Z"
  name: loadbalancer
  namespace: default
  labels:
    service.kubernetes.io/service-proxy-name: others
  resourceVersion: "1753560"
  uid: 22cf2537-4c4d-4fe5-9bbb-b8355b14d827
subsets:
- addresses:
  - ip: 100.96.1.10
    nodeName: w1-md-0-54cbf578bf-nt4d8
    targetRef:
      kind: Pod
      name: loadbalancer-78df87b854-8t844
      namespace: default
      resourceVersion: "1752571"
      uid: 388e0bad-abb0-4f47-870a-2ad1f01a428e
  - ip: 100.96.1.11
    nodeName: w1-md-0-54cbf578bf-nt4d8
    targetRef:
      kind: Pod
      name: loadbalancer-78df87b854-nkpcl
      namespace: default
      resourceVersion: "1752573"
      uid: 46cefe5a-efe8-434a-8209-2eafd717e4ee
  ports:
  - port: 80
    protocol: TCP
```
3. Verify the Service is not accessible, and the EndpointSlice has the label too
```
addressType: IPv4
apiVersion: discovery.k8s.io/v1beta1
endpoints:
- addresses:
  - 100.96.1.10
  conditions:
    ready: true
  targetRef:
    kind: Pod
    name: loadbalancer-78df87b854-8t844
    namespace: default
    resourceVersion: "1752571"
    uid: 388e0bad-abb0-4f47-870a-2ad1f01a428e
  topology:
    kubernetes.io/hostname: w1-md-0-54cbf578bf-nt4d8
- addresses:
  - 100.96.1.11
  conditions:
    ready: true
  targetRef:
    kind: Pod
    name: loadbalancer-78df87b854-nkpcl
    namespace: default
    resourceVersion: "1752573"
    uid: 46cefe5a-efe8-434a-8209-2eafd717e4ee
  topology:
    kubernetes.io/hostname: w1-md-0-54cbf578bf-nt4d8
kind: EndpointSlice
metadata:
  creationTimestamp: "2021-03-18T23:29:30Z"
  generateName: loadbalancer-
  generation: 1
  labels:
    endpointslice.kubernetes.io/managed-by: endpointslicemirroring-controller.k8s.io
    kubernetes.io/service-name: loadbalancer
    service.kubernetes.io/service-proxy-name: others
  managedFields:
  - apiVersion: discovery.k8s.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:addressType: {}
      f:endpoints: {}
      f:metadata:
        f:generateName: {}
        f:labels:
          .: {}
          f:endpointslice.kubernetes.io/managed-by: {}
          f:kubernetes.io/service-name: {}
          f:service.kubernetes.io/service-proxy-name: {}
        f:ownerReferences:
          .: {}
          k:{"uid":"22cf2537-4c4d-4fe5-9bbb-b8355b14d827"}:
            .: {}
            f:apiVersion: {}
            f:blockOwnerDeletion: {}
            f:controller: {}
            f:kind: {}
            f:name: {}
            f:uid: {}
      f:ports: {}
    manager: kube-controller-manager
    operation: Update
    time: "2021-03-18T23:29:30Z"
  name: loadbalancer-dp9pd
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Endpoints
    name: loadbalancer
    uid: 22cf2537-4c4d-4fe5-9bbb-b8355b14d827
  resourceVersion: "1753208"
  uid: 0066056b-0883-4b08-b77d-491694ecb644
ports:
- name: ""
  port: 80
  protocol: TCP
```

4. `kubectl edit` to update the Endpoint to remove the label. Found out the label is still on the EndpointSlice. In fact, I don't think updating Endpoint causes any update on the EndpointSlice

**Comment by aojea** [comment]
We fixed that only 10 days ago https://github.com/kubernetes/kubernetes/pull/98116/commits/c421e22e2c6aa4d3fbe3abd7feb0035361539d22 :)

**Comment by maplain** [comment]
Looks like it's already fixed by @aojea, and it'll be included in 1.21.

I'll test it once 21 is out of the door. Meanwhile, feel free to close this one.

**Comment by maplain** [comment]
Looks like it's already fixed by @aojea, and it'll be included in 1.21.

I'll test it once 21 is out of the door. Meanwhile, feel free to close this one. @jayunit100

**Comment by robscott** [comment]
@aojea I think it makes sense to backport this. It looks like the fix would be helpful for at least this use case, and possibly others.

**Comment by aojea** [comment]
> Looks like it's already fixed by @aojea, and it'll be included in 1.21.
> 
> I'll test it once 21 is out of the door. Meanwhile, feel free to close this one. @jayunit100

We forgot to close this :)

/close
